includes/Modules/Analytics/Advanced_Tracking/Measurement_Event_List_Factory.php

```diff
@@ -0,0 +1,103 @@
+use Google\Site_Kit\Modules\Analytics\Advanced_Tracking\Measurement_Events\Woocommerce_Event_List;
+use Google\Site_Kit\Modules\Analytics\Advanced_Tracking\Measurement_Events\WPForms_Event_List;
+use Google\Site_Kit\Modules\Analytics\Advanced_Tracking\Measurement_Events\CF7_Event_List;
+use Google\Site_Kit\Modules\Analytics\Advanced_Tracking\Measurement_Events\FormidableForms_Event_List;
+use Google\Site_Kit\Modules\Analytics\Advanced_Tracking\Measurement_Events\NinjaForms_Event_List;
+
+/**
+ * Class for detecting active plugins on a user's site and producing event lists.
+ *
+ * @since n.e.x.t.
+ * @access private
+ * @ignore
+ */
+class Measurement_Event_List_Factory {
+
+	/**
+	 * The constant check_type string for support_plugins array in the Advanced_Tracking class.
+	 *
+	 * @since n.e.x.t.
+	 * @var string
+	 */
+	const TYPE_CONSTANT = 'CONSTANT';
+
+	/**
+	 * The function check_type string for support_plugins array in the Advanced_Tracking class.
+	 *
+	 * @since n.e.x.t.
+	 * @var string
+	 */
+	const TYPE_FUNCTION = 'FUNCTION';
+
+	/**
+	 * Determines the user's current active plugins that Advanced_Tracking supports.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @param  array $supported_plugins The list of supported plugins.
+	 * @return array $active_plugins The list of active plugin event lists.
+	 */
+	public function get_active_plugin_event_lists( $supported_plugins ) {
+		$active_plugins = array_filter(
+			$supported_plugins,
+			function( $plugin_config ) {
+				if ( self::TYPE_CONSTANT === $plugin_config['check_type'] && defined( $plugin_config['check_name'] ) ) { // phpcs:ignore WordPressVIPMinimum.Constants.ConstantString.NotCheckingConstantName
+					return true;
+				}
+				if ( self::TYPE_FUNCTION === $plugin_config['check_type'] && function_exists( $plugin_config['check_name'] ) ) {
+					return true;
+				}
+				return false;
+			}
+		);
+		$plugin_names   = array_keys( $active_plugins );
+		foreach ( $plugin_names as $plugin_name ) {
+			$active_plugins[ $plugin_name ] = $this->produce_plugin_event_list( $plugin_name );
+		}
+		return $active_plugins;
+	}
+
+	/**
+	 * Instantiates an event list given a plugin name.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @param string $plugin_name The plugin's name to instantiate an event list for.
+	 * @return Measurement_Event_List|null $event_list The instantiated event list.
+	 */
+	private function produce_plugin_event_list( $plugin_name ) {
+		$event_list = null;
+		switch ( $plugin_name ) {
+			case 'WooCommerce':
+				$event_list = new Woocommerce_Event_List();
+				break;
+			case 'WPForms Lite':
+			case 'WPForms':
+				$event_list = new WPForms_Event_List();
+				break;
+			case 'Contact Form 7':
+				$event_list = new CF7_Event_List();
+				break;
+			case 'Formidable Forms':
+				$event_list = new FormidableForms_Event_List();
+				break;
+			case 'Ninja Forms':
+				$event_list = new NinjaForms_Event_List();
+				break;
+		}
+		return $event_list;
```

Comment:
Hmm, maybe we had a misunderstanding here, but I thought we were going to keep the definitions in the configuration in `Advanced_Tracking::get_supported_plugins()`. Moving this here (like it first was) scatters the plugin-based configuration into two parts. Let's keep our architecture configuration-based, there shouldn't be any plugin-specific code outside of the configuration array and the plugin's own `*_Event_List` class.

I think we should revert https://github.com/google/site-kit-wp/pull/1727/commits/1d76ad15cf2a39b2c22d5cbb692e058224b8c515, the code before was closer to that approach. The only thing we should do then is change the configuration to not instantiate the classes but provide the class names (rename `event_config_list` key to `event_list_class`); the classes can then simply be instantiated by `Advanced_Tracking::configure_events`.